src/connection_pool.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use std::{
+    collections::BTreeMap,
+    net::SocketAddr,
+    sync::{Arc, Mutex},
+};
+
+// Pool for keeping open connections. Pooled connections are associated with a `ConnectionHandle`
+// which can be used to remove them from the pool.
+#[derive(Clone)]
+pub(crate) struct ConnectionPool {
+    store: Arc<Mutex<Store>>,
+}
+
+impl ConnectionPool {
+    pub fn new() -> Self {
+        Self {
+            store: Arc::new(Mutex::new(Store::default())),
+        }
+    }
+
+    pub fn insert(&self, addr: SocketAddr, conn: quinn::Connection) -> ConnectionHandle {
+        // NOTE: this will panic if some other thread panicked while holding this lock. That's
+        // probably the correct thing to do as it avoids working with potentially invalid state
+        // (see the rust docs about mutex poisoning for more info).
+        // That said, it's highly unlikely this will ever panic as none of the operations we
+        // perform anywhere while holding the lock can panic.
+        // This comment applies to all the other instances of `unwrap` in this module.
+        let mut store = self.store.lock().unwrap();
```

Comment:
Shouldn't we return `Result` still? so if one day someone introduces a bug in some other part of the code which opens up the possibility of a panic then the user app gets an error and can handle it or report it properly

Reply:
Yes, let's get rid of unwrap in all of our code with no debates about it. Even locks can poison and if they do we can still get the data from the thread if we want and so on, so possible to handle or just return result as well. We have a policy, no unwraps in production so let's Engineer our code to not have them.

Reply:
I think calling `unwrap` on the result of `Mutex::lock` is one of the few legitimate uses of `unwrap` in production code. The reason is that `lock` returns error only if the mutex is poisoned which means some other thread panicked while holding the lock. So if this `unwrap` panics, that means there already is a panic somewhere else in our code which means our goal of never panicking in production has already failed :smiling_imp: Also, if you look at the other places where we lock this mutex and the things we do while it's being locked then you'll see that nothing there can actually panic. That means that this `unwrap` never panics either. So propagating this error upwards would just make the API annoying to use for no good reason.

Reply:
There will only be a panic elsewhere if there is code that panics in there. So allowing panic is how that happens. 
If the lock panicked we can still get the data by error handling but should probably pass back an error we cannot handle or are unlikely to. 
Safe to say code can panic in places where we `unwrap` and cannot where we don't. 
We will need to check dependencies at some stage to make sure they don't cause panics anyway as that is another wy to poison a thread we think can never be poisoned, it's worse though if we have panics in our code as well. 
IMO it's so much simpler to audit code that does not have the world panic/unwrap in it as then I need to start looking at the mem internals or thread poisoning or some other thing and the logic gets lost. 
So I would say panic nowhere and we stand a chance. 
Atm we cannot tell if some dependent crate will panic as we need to audit them, until that time these threads can poison. Even then we can still handle it with restructuring the error and deciding what to do there.

Reply:
> Also, if you look at the other places where we lock this mutex and the things we do while it's being locked then you'll see that nothing there can actually panic. That means that this unwrap never panics either.
This is what I was trying to address in my previous comment, in the future someone may add/change code in those places which start panicking, specially if they see we use panic in some places like here, so by having any unwrap we implicitly say "we allow our code to panic sometimes" then new devs will start adding unwraps which will break the invariant you are referring to.

Reply:
OK guys, will change this to use `PoisonError::into_inner` just to not waste time on this even though I disagree with you on this one.

Reply:
Also @madadam  Can you put a TODO here 
//TODO Check data plus invariants here!!

Reply:
How about doing that thing now instead of TODO it? But then I'm not sure what you mean by "check data plus invariants". Could you elaborate?

Reply:
Yea no worries, not in all cases would this be an error we cannot "fix" (this is why we don't want panic by default). 
So in these cases we can look at what was in the thread behind the guard and if there is data we need, then we can still access it, if the invariants all still hold then we can continue (even if we need to do some other work), otherwise we would pass this back and see if anyone else can handle it. If not it's all the way back to main and fails (iff invariants are broken)